HealthcareSystem/Backend/Service/ExaminationAndPatientCard/ExaminationService.cs

```diff
@@ -28,6 +29,336 @@ public void AddExamination(Examination examination)
             _scheduledExaminationRepository.AddExamination(examination);
         }
 
+        public List<Examination> AdvancedSearchExaminationReports(List<Examination> examinations, string startDate, string endDate, string doctorSurname, string anamesis, string firstOperator, string secondOperator, string thirdOperator)
```

Comment:
Ok, so, this is the main thing to fix, since it's obviously not very readable. My suggestion here would be to use a modified version of the [Specification Pattern](https://www.codeproject.com/Articles/670115/Specification-pattern-in-Csharp). The modifications I suggest is to skip the Not specification since you're not using it anyway, to make one LogicalOperation method instead of And and Or to make it easier to use with the exact data you're getting, and to skip the ExpressionSpecification and make specific classes for what you need instead. 

So, here is approximately how I suggest implementing the advanced search specification for examinations, and how it would be used (with the assumption that the DTO suggestion is implemented as I said, adjust accordingly for changes in that). One small issue, I forgot to handle what happens if some of the operators is null, but I don't think that can actually happen.

###  Specification
```
    public interface ISpecification<T>
    {
        bool IsSatisfiedBy(T o);
        ISpecification<T> BinaryOperation(LogicalOperator op, ISpecification<T> specification);
    }
```
```
    public abstract class CompositeSpecification<T> : ISpecification<T>
    {
        public ISpecification<T> BinaryOperation(LogicalOperator op, ISpecification<T> specification)
        {
            switch (op)
            {
                case LogicalOperator.And:
                    return new AndSpecification<T>(this, specification);
                case LogicalOperator.Or:
                    return new OrSpecification<T>(this, specification);
                default:
                    return this;
            }
        }

        public abstract bool IsSatisfiedBy(T o);
    }
```
```
    public class AndSpecification<T> : CompositeSpecification<T>
    {
        ISpecification<T> _leftSpecification;
        ISpecification<T> _rightSpecification;

        public AndSpecification(ISpecification<T> left, ISpecification<T> right)
        {
            _leftSpecification = left;
            _rightSpecification = right;
        }

        public override bool IsSatisfiedBy(T o)
        {
            return _leftSpecification.IsSatisfiedBy(o)
                && _rightSpecification.IsSatisfiedBy(o);
        }
    }
```
```
   public class OrSpecification<T> : CompositeSpecification<T>
    {
        ISpecification<T> _leftSpecification;
        ISpecification<T> _rightSpecification;

        public OrSpecification(ISpecification<T> left, ISpecification<T> right)
        {
            _leftSpecification = left;
            _rightSpecification = right;
        }

        public override bool IsSatisfiedBy(T o)
        {
            return _leftSpecification.IsSatisfiedBy(o)
                || _rightSpecification.IsSatisfiedBy(o);
        }
    }
```
```
    public class StartDateSpecification : CompositeSpecification<Examination>
    {
        private DateTime? _startDate;

        public StartDateSpecification(DateTime? startDate)
        {
            _startDate = startDate;
        }

        public override bool IsSatisfiedBy(Examination o)
        {
            if (_startDate is null)
                return true;

            // TODO implement whatever check actually needs to happen here,
            // i'm not sure what start date is meant to be filtering for
            return true;
        }
    }
```
```
    public class EndDateSpecification : CompositeSpecification<Examination>
    {
        private DateTime? _endDate;

        public EndDateSpecification(DateTime? endDate)
        {
            _endDate = endDate;
        }

        public override bool IsSatisfiedBy(Examination o)
        {
            if (_endDate is null)
                return true;

            // TODO implement whatever check actually needs to happen here,
            // i'm not sure what end date is meant to be filtering for
            return true;
        }
    }
```
```
    public class DoctorSurnameSpecification : CompositeSpecification<Examination>
    {
        private string _doctorSurname;

        public DoctorSurnameSpecification(string doctorSurname)
        {
            _doctorSurname = doctorSurname;
        }

        public override bool IsSatisfiedBy(Examination o)
        {
            if (_doctorSurname is null)
                return true;

            return _doctorSurname.Equals(o.Doctor.Surname);
        }
    }
```
```
   public class AnamnesisSpecification : CompositeSpecification<Examination>
    {
        private string _anamnesis;

        public AnamnesisSpecification(string anamnesis)
        {
            _anamnesis = anamnesis;
        }

        public override bool IsSatisfiedBy(Examination o)
        {
            if (_anamnesis is null)
                return true;

            return _anamnesis.Equals(o.Anamnesis);
        }
    }
```
### Usage
```
public class ExaminationSearchDTO
    {
        public DateTime? StartDate { get; set; }
        public LogicalOperator EndDateOperator { get; set; }
        public DateTime? EndDate { get; set; }
        public LogicalOperator AnamnesisOperator { get; set; }
        public string Anamnesis { get; set; }
        public LogicalOperator DoctorSurnameOperator { get; set; }
        public string DoctorSurname { get; set; }
    }
```
```
        public IEnumerable<Examination> AdvancedSearch(string jmbg, ExaminationSearchDTO parameters)
        {
            List<Examination> examinations = repo.GetExaminationsByPatient(jmbg);

            ISpecification<Examination> filter = new StartDateSpecification(parameters.StartDate);
            filter = filter.BinaryOperation(parameters.EndDateOperator, new EndDateSpecification(parameters.EndDate));
            filter = filter.BinaryOperation(parameters.AnamnesisOperator, new AnamnesisSpecification(parameters.Anamnesis));
            filter = filter.BinaryOperation(parameters.DoctorSurnameOperator, new DoctorSurnameSpecification(parameters.DoctorSurname));

            return examinations.Where(examination => filter.IsSatisfiedBy(examination));
        }
```
### Why bother with this?
* Individual filters are easy to test. The full specification is also easy to test on an individual examination.
* Easy to change in case your search parameters change. What would you do with your current method if you were asked to add a parameter (other than cry, pretty much)? As for what you would do here - maybe add one more specification class, instantiate accordingly, connect to existing parameters any way you like.
### Additional suggestions
* Maybe the object that holds the search parameters could be a bit smarter than a DTO and instantiate the specification object. 
* @PSW-2020-ORG4/devops Any other suggestions to make this better maybe? I wish it was simpler, but this really just isn't a simple functionality.

Reply:
Applied in a6a4faba184bae4fbbe467acddb31e9e492749d9.